### ISSUE TITLE
cargo test incorrectly warns for dead code

### ISSUE THREAD

**Issue by jean553** [rust-lang/rust#46379]
Hi!

I got unexpected dead code warnings when executing `cargo test` with multiple test files. I use a very simple example below that reproduces the issue.

I have two files that contains tests, `tests/test_one.rs` and `tests/test_two.rs`. Both contains exactly the same content (except for the unique test function name they contain):

```rust
mod routines;

#[test]
fn test_one() {
    routines::my_routine();
}
```

And another file called `tests/routines.rs` that simply contains:

```rust
pub fn my_routine() {
}
```

When I execute `cargo test`, the two tests are executed successfully and there is no raised warning. But if I remove the `my_routine()` call from one of the two tests, `cargo test` stills end up successfully but raises a warning on `pub fn routine()` saying `function is never used`. However, one test still calls the function, so there is no dead code as stated. 

I got the same issue with both rust stable (1.22.1) and rust nightly (1.24.0).

Thanks.

**Comment by steveklabnik** [comment]
This is not actually a bug, as far as I know!

>And another file called tests/routines.rs that simply contains:

Cargo is *also* attempting to compile this file as its own test. See the explanation here: https://doc.rust-lang.org/book/second-edition/ch11-03-test-organization.html#submodules-in-integration-tests

I believe if you move this file to `tests/routines/mod.rs`, this will work without warning.

**Comment by jean553** [comment]
Yes, when replacing `tests/routines.rs` by `tests/routines/mod.rs`, Cargo does not attempt to compile `routines.rs` as a test.

But, the warning is still raised at the first execution of `cargo test` (and only at the first execution).

Here the details of what I do after changing `tests/routines.rs` to `tests/routines/mod.rs`:
1. execute `cargo test` multiple times with two calls to `my_routine()` (one in each test) -> no warning
2. remove one call to `my_routine()` from one of the two tests
3. execute `cargo test` -> one warning `function is never used` is raised for `my_routine()`
4. execute `cargo test` again (without code modification) -> no warning anymore

For exactly the same code, `cargo test` outputs a warning only during the first execution. Furthermore, for each execution, the two tests are run so `my_routine()` is called. Maybe I just miss something, but this is kind of weird, isn't it ?

**Comment by nickolay** [comment]
I believe cargo compiles each "integration test" (i.e. file in `tests/*.rs`) independently. So if the helper in a common module `tests/routines/mod.rs` is used in `tests/test_one.rs`, but not in `tests/test_two.rs`, there will be a warning when compiling `test_two.rs`.

(`cargo test` re-builds only what needs to be rebuilt, that's why it only warns on the first execution.)

I've worked around this by using `mod routines; pub use routines::*;` in `tests/*.rs`.

**Comment by svenstaro** [comment]
So we're running into the same issue and I think the workaround `mod routines; pub use routines::*;` is not optimal as that does hide truly dead code. I think this is a legitimate cargo bug.

**Comment by KSXGitHub** [comment]
This is either a bug or a design flaw.

In TypeScript, an exported item (even if it is never used) is never considered dead code. I don't why it is in Rust.

**Comment by nickolay** [comment]
@KSXGitHub well, if you export the items from the crate with `mod routines; pub use routines::*;` it's not considered dead code.

What @svenstaro is asking for would require considering all targets/configurations before determining if an item is dead code instead of letting the rust compiler report them when compiling an individual crate...

**Comment by jweinst1** [comment]
I am noticing this issue,

when will the PR above be released in rust ? ^

**Comment by eftychis** [comment]
We are encountering the same issue. I am not sure I am comfortable with no dead code warnings. Perhaps an alternative approach to organizing code people would suggest if there is no re-design any time soon?

**Comment by pradovic** [comment]
Actually, just using `pub mod routines;`, in every test file that I wish to use `routines` module solved the problem for me. In the files that don't use `routines`, I just omit declaring the module at all and there are no warnings, if I am not mistaken. This makes sense to me, as I don't see the need of an unused warning for not using all the pub functions of another module.

**Comment by bachrc** [comment]
We are still encountering this issue with the last version of Rust (1.45.0 nightly)

**Comment by bhgomes** [comment]
Any updates on this? Still encountering this issue on `v1.58.0-nightly`.

**Comment by X01XX** [comment]
Allowing dead code warnings, and running "cargo test", seems to show truly dead code warnings.

**Comment by xixixao** [comment]
I can still reproduce this with `trait` and `impl`.

```rs
# tests/utils/mod.rs
pub struct Foo {}
impl Foo {
    pub fn bar(self) -> Self {
        println!("hello!");
        self
    }
}
```


```rs
# tests/some_test.rs
mod utils;

use crate::utils::{Foo};

#[test]
fn baz() {
    (Foo {}).bar();
}
```
This warns about `Foo` and `bar` being dead code.

Note that `pub fn`s  in `tests/utils/mod.rs` do not suffer from this.

**Comment by ik1ne** [comment]
The rust book's [11.3. Test Organization](https://doc.rust-lang.org/book/ch11-03-test-organization.html) encourages to create `tests/common/mod.rs` for test setup codes.

However, this results to false positive in some cases such as `common::common_func1()` is only used by `tests/test1.rs` and `common::common_func2()` is only used by `tests/test2.rs`.

**Comment by CobaltCause** [comment]
A workaround for this issue is to use the `tests/it/main.rs` pattern as described by <https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html> (although I usually spell it `tests/integrations/main.rs` in my own projects).

**Comment by ClementTsang** [comment]
+1 to following the `tests/it/main.rs` pattern in https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html as a workaround.

**Comment by daniel-wong-dfinity-org** [comment]
I think that while this is very understandably surprising, this is TECHNICALLY not a bug. I think the key facts here are that

1. tests/test_one.rs and tests/test_two.rs are compiled into SEPARATE crates, but
2. the compilations HAPPEN to both share a common source file, namely, tests/routines.rs.

Therefore, even though you only have one routines.rs file, you are defining my_routine TWICE: once when you compile test_one and again when you compile test_two. If nothing within test_two calls my_routine, and since my_routine is in a private (sub)module, there is no hope that this second definition could ever be used, and so this second definition is "dead".

**_TANGENT_**: IMHO, using the same source file (tests/routines.rs) is the "original sin" here, and there ought to be a warning, but I'm not sure how this would be detected. This should be avoided because it leads to this sort of confusion (wait, this one piece of code gives rise to TWO SEPARATE definitions??). But if you insist on using the same source file in the compilation of separate crates (because a book indicated that this is the "right way" to do things), well, you can avoid this warning by sprinkling the `pub` keyword into your code.

The question is NOT, "Is this CODE used anywhere?". The question is, "Is this DEFINITION used anywhere?". And again, what you must remember is that there are TWO definitions of my_routine. One definition is used (by the crate that it is part of), but the other definition is not used.

But yeah, this behavior is not fun.
